src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java

```diff
@@ -145,11 +163,20 @@ public FileSystemSecurityRealm(final Path root, final int levels) {
      * @param root the root path of the identity store
      */
     public FileSystemSecurityRealm(final Path root) {
-        this(root, NameRewriter.IDENTITY_REWRITER, 2);
+        this(root, NameRewriter.IDENTITY_REWRITER, 2, true);
     }
 
-    private Path pathFor(final String name) {
+    private Path pathFor(String name) {
         assert name.codePointCount(0, name.length()) > 0;
+        if (encoded) {
+            String normalized = Normalizer.normalize(name.toLowerCase(Locale.ROOT), Normalizer.Form.NFKC)
+                    .replaceAll("[^a-z0-9]", "_");
+
+            String base32 = ByteIterator.ofBytes(new ByteStringBuilder().append(name).toArray())
+                    .base32Encode(Alphabet.Base32Alphabet.STANDARD, false).drainToString();
+
+            name = normalized + "-" + base32;
+        }
         final int levels = this.levels;
         Path path = root;
         int idx = 0;
```

Comment:
I think this could behave strangely if the normalized/folded name is less than the number of hash levels.  To avoid problems, the name should be fixed first, then hash levels done, then base32-encoding added before ```.xml```.